app/src/main/java/ch/epfl/sdp/ui/main/AuthFragment.java

```diff
@@ -0,0 +1,124 @@
+package ch.epfl.sdp.ui.main;
+
+import android.content.Intent;
+import android.os.Bundle;
+import android.util.Log;
+import android.view.LayoutInflater;
+import android.view.View;
+import android.view.ViewGroup;
+
+import com.google.android.gms.auth.api.signin.GoogleSignIn;
+import com.google.android.gms.auth.api.signin.GoogleSignInAccount;
+import com.google.android.gms.auth.api.signin.GoogleSignInClient;
+import com.google.android.gms.auth.api.signin.GoogleSignInOptions;
+import com.google.android.gms.common.api.ApiException;
+import com.google.android.gms.tasks.Task;
+import com.google.firebase.auth.AuthCredential;
+import com.google.firebase.auth.FirebaseAuth;
+import com.google.firebase.auth.GoogleAuthProvider;
+
+import androidx.annotation.NonNull;
+import androidx.annotation.Nullable;
+import androidx.fragment.app.Fragment;
+import androidx.lifecycle.ViewModelProvider;
+import ch.epfl.sdp.auth.firebase.FirebaseAuthenticator;
+import ch.epfl.sdp.R;
+import ch.epfl.sdp.databinding.AuthFragmentBinding;
+
+public class AuthFragment extends Fragment implements View.OnClickListener {
+
+    private final static String TAG = "AuthFragment";
+    private final static int RC_GOOGLE_SIGN_IN = 9001;
+
+    private AuthViewModel mViewModel;
+    private AuthFragmentBinding mBinding;
+
+    private GoogleSignInClient mGoogleSignInClient;
+    private FirebaseAuthenticator mAuthenticator;
```

Comment:
This defeats the purpose of building an abstraction: you can't replace this with a mock implementation when testing, as the required type is FirebaseAuthenticator. Use Authenticator instead :)

Reply:
The thing is that Google Sign-In and other authentication services return AuthCredentials which is purely a Firebase thing. So in this case I need to make it a FirebaseAuthenticator as the interface does not support the AuthCredential type. The solution would be to create a Credential class that wraps the AuthCredential.